jscomp/others/js_array.ml

```diff
@@ -17,61 +17,122 @@
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU Lesser General Public License for more details.
- * 
+ *
  * You should have received a copy of the GNU Lesser General Public License
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)
 
-
 type 'a t = 'a array
+type 'a array_like
+type 'a array_iter = 'a array_like (* don't think this is very useful to implement wihtout language support *)
+
+external make : int -> unit Js.undefined array = "Array" [@@bs.new]
+
+external from : 'a array_like -> 'b array = "Array.from" [@@bs.val] (* es2015 *)
+external unsafeFrom : 'a -> 'b array = "Array.from" [@@bs.val] (* es2015 *)
+external fromMap : 'a array_like -> ('a -> 'b [@bs]) -> 'b array = "Array.from" [@@bs.val] (* es2015 *)
+external unsafeFromMap : 'a -> ('b -> 'c [@bs]) -> 'c array = "Array.from" [@@bs.val] (* es2015 *)
+external isArray : 'a -> Js.boolean = "Array.isArray" [@@bs.val] (* es2015 *)
+(* Array.of: seems pointless unless you can bind *) (* es2015 *)
+
+external length : 'a array -> int = "" [@@bs.get]
+
+
+(* Mutator functions
+*)
+external copyWithin : to_:int -> 'this = "" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
+external copyWithinFrom : to_:int -> from:int -> 'this = "copyWithin" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
+external copyWithinFromRange : to_:int -> start:int -> end_:int -> 'this = "copyWithin" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
+
+external fill : 'a -> 'this = "" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
```

Comment:
I agree with making imperative functions explicit, but this specific case seems pretty explicit to me. I don't know what else it could mean. Unless we employ a more generic convention like `fillMut`, I don't know what I'd change this to.